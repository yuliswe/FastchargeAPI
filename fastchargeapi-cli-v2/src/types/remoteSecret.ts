export type CliRemoteSercretResult = {
  status: "success" | "canceled";
};
