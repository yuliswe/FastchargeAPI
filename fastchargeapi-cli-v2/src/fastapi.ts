import { createFastapiProgram } from "src/fastapi/program";

const program = createFastapiProgram();
program.parse(process.argv);
