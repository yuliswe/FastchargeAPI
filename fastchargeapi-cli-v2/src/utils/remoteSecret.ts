import crypto from "crypto";
import { EncryptJWT, JWTPayload, SignJWT, jwtDecrypt, jwtVerify } from "jose";
import { graphql } from "src/__generated__/gql";
import { getGQLClient } from "src/graphqlClient";

export async function getRemoteSecret(args: {
  key: string;
  jweSecret: Uint8Array;
  jwtSecret: Uint8Array;
}): Promise<object> {
  const { key, jweSecret, jwtSecret } = args;
  const resp = await getGQLClient().query({
    query: graphql(`
      query GetSecret($key: String!) {
        getSecret(key: $key) {
          value
        }
      }
    `),
    variables: { key },
  });

  const secret = resp.data.getSecret.value;

  const jwtVerified = await jwtVerify(secret, jwtSecret, { algorithms: ["HS512"] });

  const { encrypted } = jwtVerified.payload as { encrypted: string };
  const jweDecrypted = await jwtDecrypt(encrypted, jweSecret);
  return jweDecrypted.payload;
}

export async function setRemoteSecret(args: {
  key: string;
  value: JWTPayload;
  description?: string;
  expireAt?: number;
  jweSecret: Uint8Array;
  jwtSecret: Uint8Array;
}) {
  const { key, value, description, expireAt, jweSecret, jwtSecret } = args;
  const signedValue = await encryptAndSign(value, { jweSecret, jwtSecret });
  const response = await getGQLClient().mutate({
    mutation: graphql(`
      mutation PutSecret($key: String!, $signedValue: String!, $description: String, $expireAt: Timestamp) {
        createSecret(key: $key, value: $signedValue, description: $description, expireAt: $expireAt) {
          createdAt
        }
      }
    `),
    variables: {
      key,
      signedValue,
      description,
      expireAt,
    },
  });
  return response;
}

export async function encryptAndSign(
  body: JWTPayload,
  options: {
    jweSecret: Uint8Array;
    jwtSecret: Uint8Array;
  }
): Promise<string> {
  const { jweSecret, jwtSecret } = options;
  const encrypted = await new EncryptJWT(body)
    .setIssuedAt()
    .setIssuer("fastchargeapi.com")
    .setAudience("fastchargeapi.com")
    .setProtectedHeader({
      alg: "dir",
      enc: "A256CBC-HS512",
    })
    .encrypt(jweSecret);

  const signed = await new SignJWT({
    encrypted,
  })
    .setProtectedHeader({
      alg: "HS512",
    })
    .sign(jwtSecret);
  return signed;
}

export function createSecret(): Uint8Array {
  const secret = new Uint8Array(64);
  crypto.getRandomValues(secret);
  return secret;
}

export function createSecretPair(): { jweSecret: Uint8Array; jwtSecret: Uint8Array } {
  return { jweSecret: createSecret(), jwtSecret: createSecret() };
}
