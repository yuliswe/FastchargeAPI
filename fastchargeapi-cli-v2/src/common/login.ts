import chalk from "chalk";
import { Command, program } from "commander";
import opener from "opener";
import { graphql } from "src/__generated__/gql";
import { envVars, reactHost } from "src/env";
import { getGQLClient } from "src/graphqlClient";
import { NotFoundSimpleGQLError } from "src/simplifiedGQLErrors";
import { tiChecker } from "src/tiChecker";
import { CliCommonLoginCommandOptions, CliGlobalOptions } from "src/types/cliOptions";
import { createSecretPair, getRemoteSecret } from "src/utils/remoteSecret";
import * as uuid from "uuid";
import { readOrRefreshAuthFile, verifyIdToken, writeToAuthFile } from "../utils/authFile";

export const createCommonLoginCommand = () =>
  new Command("login")
    .summary("Login as a user")
    .description("Stores the user credentials in the system.")
    .helpOption("--help", "Display help for command")
    .option("--profile <char>", "Store the user credentials in a profile")
    .action(async (options) => {
      await loginCommand({
        globalOptions: tiChecker.CliGlobalOptions.strictFrom(program.opts()),
        options: tiChecker.CliCommonLoginCommandOptions.strictFrom(options),
      });
    });

export const loginCommand = async (args?: {
  globalOptions?: CliGlobalOptions;
  options?: CliCommonLoginCommandOptions;
}) => {
  const { globalOptions, options } = args ?? {};
  const profile = globalOptions?.profile ?? options?.profile ?? undefined;
  try {
    const auth = await readOrRefreshAuthFile({ profile });
    if (auth) {
      console.log(profile ? `Login successful for profile '${profile}'.` : "Login successful.");
      if (envVars.SHOW_AUTH) {
        console.log(auth);
      }
      return auth;
    }
  } catch (e) {
    // ignore error and proceed to login
  }

  const key = uuid.v4();
  const { jweSecret, jwtSecret } = createSecretPair();
  const hex = (arr: Uint8Array) =>
    Array.from(arr)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  const params = new URLSearchParams({
    relogin: "true",
    behavior: "putsecret",
    jwe: hex(jweSecret),
    jwt: hex(jwtSecret),
    key,
  });
  const url = new URL(`/auth/?${params.toString()}`, reactHost);
  opener(url.href);
  console.log(chalk.yellow("Please authenticate in the browser."));
  console.log("If the browser does not open, please visit:");
  console.log(chalk.blue(url.href));

  let idToken = "";
  let refreshToken = "";
  let tries = 0;
  while (tries < 120) {
    await new Promise((r) => setTimeout(r, 1000));
    tries += 1;
    try {
      const value = await getRemoteSecret({ key, jweSecret, jwtSecret });
      ({ idToken, refreshToken } = tiChecker.RefreshIdTokenResult.from(value));
      break;
    } catch (e) {
      if (e instanceof NotFoundSimpleGQLError) {
        if (tries >= 40) {
          console.log("Timed out. Press enter to retry.");
          await new Promise((r) => setTimeout(r, 3000));
        }
        continue;
      }
      throw e;
    }
  }
  const { email } = await verifyIdToken(idToken);
  if (!email) {
    throw new Error("ID token is invalid. Login failed.");
  }
  const resp = await getGQLClient({ email, idToken }).query({
    query: graphql(`
      query GetUserPKByEmail($email: Email!) {
        getUserByEmail(email: $email) {
          pk
        }
      }
    `),
    variables: { email },
  });
  const userPK = resp.data.getUserByEmail.pk;
  const authFileContent = await writeToAuthFile(profile, { idToken, refreshToken, userPK, email });
  if (envVars.SHOW_AUTH) {
    const auth = await readOrRefreshAuthFile({ profile }); // populates user pk and email
    console.log(profile ? `Login successful for profile '${profile}'.` : "Login successful.");
    console.log(auth);
  }

  console.log(chalk.green("Login successful."));
  return authFileContent;
};
