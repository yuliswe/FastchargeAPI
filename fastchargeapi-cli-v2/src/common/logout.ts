import { Command, program } from "commander";
import fs from "fs/promises";
import { tiChecker } from "src/tiChecker";
import { CliCommonLogoutCommandOptions, CliGlobalOptions } from "src/types/cliOptions";
import { deleteAuthFile, listAuthFiles } from "../utils/authFile";

export const createCommonLogoutCommand = () =>
  new Command("logout")
    .summary("Logout a user")
    .description("Clears the user credentials from the system.")
    .helpOption("--help", "Display help for command")
    .option("--profile <char>", "Clear the user credentials for a specific profile")
    .action(async (options) => {
      await logoutCommand({
        globalOptions: tiChecker.CliGlobalOptions.strictFrom(program.opts()),
        options: tiChecker.CliCommonLogoutCommandOptions.strictFrom(options),
      });
    });

export async function logoutCommand(args?: {
  globalOptions: CliGlobalOptions;
  options: CliCommonLogoutCommandOptions;
}) {
  const { globalOptions, options } = args ?? {};
  const profile = globalOptions?.profile ?? options?.profile ?? undefined;
  if (profile) {
    await deleteAuthFile(profile);
  } else {
    for (const file of await listAuthFiles()) {
      await fs.unlink(file);
    }
  }
  console.log("Logged out.");
}
