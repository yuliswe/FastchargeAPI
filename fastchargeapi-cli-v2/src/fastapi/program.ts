import { Command } from "commander";
import { createCommonLoginCommand } from "src/common/login";
import { createCommonLogoutCommand } from "src/common/logout";

export const createFastapiProgram = () =>
  new Command("fastapi")
    .option("--profile <char>", "Run command as a specific user")
    .usage("[global-options] command")
    .version("0.0.1")
    .description("fastapi cli")
    .addHelpCommand(false)
    .helpOption("--help", "Display help for command")
    .addCommand(createCommonLoginCommand())
    .addCommand(createCommonLogoutCommand());
