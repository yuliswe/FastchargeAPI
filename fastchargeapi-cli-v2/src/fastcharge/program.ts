import { Command } from "commander";
import { createCommonLoginCommand } from "src/common/login";
import { createCommonLogoutCommand } from "src/common/logout";

export const createFastchargeProgram = () =>
  new Command("fastcharge")
    .version("0.0.1")
    .description("fastcharge cli")
    .usage("[global-options] command")
    .helpOption("--help", "Display help for command")
    .addHelpCommand(false)
    .option("--profile <char>", "run command as a specific user")
    .addCommand(createCommonLoginCommand())
    .addCommand(createCommonLogoutCommand());
