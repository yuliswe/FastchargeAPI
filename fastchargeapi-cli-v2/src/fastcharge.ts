import { createFastchargeProgram } from "src/fastcharge/program";

const program = createFastchargeProgram();
program.parse(process.argv);
