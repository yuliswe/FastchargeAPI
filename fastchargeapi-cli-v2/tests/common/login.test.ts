import * as loginModule from "src/common/login";
import { createFastapiProgram } from "src/fastapi/program";
import { createFastchargeProgram } from "src/fastcharge/program";
import * as remoteSecretModule from "src/utils/remoteSecret";
import { createRandomHex, createSecret, hex, setRemoteSecret, waitForSecretContent } from "src/utils/remoteSecret";
import { ConsoleLogSpy, createConsoleLogSpy, fastapi, fastcharge, getAdminUserCredentials } from "tests/utils";

jest.mock("opener", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("fastcharge login --help", () => {
  it("prints help message", async () => {
    const { stdout, exitCode } = await fastcharge(["login", "--help"]);
    expect(exitCode).toBe(0);
    expect(stdout.getOutput()).toMatchSnapshot();
  });
});

describe("fastapi login --help", () => {
  it("prints help message", async () => {
    const { stdout, exitCode } = await fastapi(["login", "--help"]);
    expect(exitCode).toBe(0);
    expect(stdout.getOutput()).toMatchSnapshot();
  });
});

describe("loginCommand", () => {
  let consoleLogSpy: ConsoleLogSpy;

  beforeEach(() => {
    consoleLogSpy = createConsoleLogSpy();
  });

  for (const profile of [undefined, "myprofile"]) {
    for (const program of [createFastapiProgram(), createFastchargeProgram()]) {
      // eslint-disable-next-line jest/valid-title
      it(`${program.name()} login` + (profile ? ` --profile ${profile}` : ""), async () => {
        const jweSecret = createSecret();
        const jwtSecret = createSecret();
        const key = createRandomHex({ nchars: 32 });
        const adminUserCredentials = await getAdminUserCredentials();
        // This would be set by the web app after user is logged in.
        await setRemoteSecret({
          key,
          value: adminUserCredentials,
          jweSecret,
          jwtSecret,
        });
        const originalWaitForSecretContent = waitForSecretContent;
        jest.spyOn(remoteSecretModule, "waitForSecretContent").mockImplementation((args) => {
          return originalWaitForSecretContent({
            ...args,
            key,
            jweSecret,
            jwtSecret,
          });
        });
        const loginCommandSpy = jest.spyOn(loginModule, "loginCommand");
        const openerModule = jest.requireMock("opener");
        const opener = jest.spyOn(openerModule, "default");
        const args = ["login"];
        if (profile) {
          args.push("--profile", profile);
        }
        await program.parseAsync(args, { from: "user" });
        const url = new URL(opener.mock.lastCall?.[0] as string);
        const params = Object.fromEntries(url.searchParams.entries());
        expect(params).toMatchSnapshotExceptForProps({
          jwe: hex(jweSecret),
          jwt: hex(jwtSecret),
          key,
        });
        expect(loginCommandSpy.mock.lastCall).toMatchSnapshot(
          `loginCommand must be called with the right profile name: ${profile}`
        );
        const authFileContent = await loginCommandSpy.mock.results[0].value;
        expect(authFileContent).toMatchSnapshotExceptForProps({
          idToken: adminUserCredentials.idToken,
          refreshToken: adminUserCredentials.refreshToken,
          email: expect.any(String),
          userPK: expect.any(String),
        });
        expect(
          consoleLogSpy.getOutput({
            redactLine: { URL: (ln) => ln.includes("https://") },
          })
        ).toMatchSnapshot("console output");
      });
    }
  }
});
