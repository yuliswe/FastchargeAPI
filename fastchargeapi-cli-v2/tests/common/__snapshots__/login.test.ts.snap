// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginCommand fastapi login --profile myprofile 1`] = `
{
  "behavior": "putsecret",
  "jwe": Any<String>,
  "jwt": Any<String>,
  "key": Any<String>,
  "relogin": "true",
}
`;

exports[`loginCommand fastapi login --profile myprofile 2`] = `
{
  "email": Any<String>,
  "idToken": Any<String>,
  "issuer": "firebase",
  "refreshToken": Any<String>,
  "userPK": Any<String>,
}
`;

exports[`loginCommand fastapi login --profile myprofile: console output 1`] = `
"[33mPlease authenticate in the browser.[39m
If the browser does not open, please visit:
<REDACTED:URL>
[32mLogin successful.[39m
"
`;

exports[`loginCommand fastapi login 1`] = `
{
  "behavior": "putsecret",
  "jwe": Any<String>,
  "jwt": Any<String>,
  "key": Any<String>,
  "relogin": "true",
}
`;

exports[`loginCommand fastapi login 2`] = `
{
  "email": Any<String>,
  "idToken": Any<String>,
  "issuer": "firebase",
  "refreshToken": Any<String>,
  "userPK": Any<String>,
}
`;

exports[`loginCommand fastapi login: console output 1`] = `
"[33mPlease authenticate in the browser.[39m
If the browser does not open, please visit:
<REDACTED:URL>
[32mLogin successful.[39m
"
`;

exports[`loginCommand fastcharge login --profile myprofile 1`] = `
{
  "behavior": "putsecret",
  "jwe": Any<String>,
  "jwt": Any<String>,
  "key": Any<String>,
  "relogin": "true",
}
`;

exports[`loginCommand fastcharge login --profile myprofile 2`] = `
{
  "email": Any<String>,
  "idToken": Any<String>,
  "issuer": "firebase",
  "refreshToken": Any<String>,
  "userPK": Any<String>,
}
`;

exports[`loginCommand fastcharge login --profile myprofile: console output 1`] = `
"[33mPlease authenticate in the browser.[39m
If the browser does not open, please visit:
<REDACTED:URL>
[32mLogin successful.[39m
"
`;

exports[`loginCommand fastcharge login 1`] = `
{
  "behavior": "putsecret",
  "jwe": Any<String>,
  "jwt": Any<String>,
  "key": Any<String>,
  "relogin": "true",
}
`;

exports[`loginCommand fastcharge login 2`] = `
{
  "email": Any<String>,
  "idToken": Any<String>,
  "issuer": "firebase",
  "refreshToken": Any<String>,
  "userPK": Any<String>,
}
`;

exports[`loginCommand fastcharge login: console output 1`] = `
"[33mPlease authenticate in the browser.[39m
If the browser does not open, please visit:
<REDACTED:URL>
[32mLogin successful.[39m
"
`;
